### PR TITLE
Added a method to PARStore for fetching change predecessors.

### PR DESCRIPTION
**Pull request by drewmccormack**
It fetches the most recent predecessor of each change, and returns in a dictionary.

### CODE REVIEW

**Comment by cparnot** [comment]
@drewmccormack Looking good to me. I had to look at the code again for `fetchChangesMatchingPredicate:deviceIdentifier` before I realized the result is sorted by timestamp, which makes it easy to reason about the order of the returned changes. I also wondered about the order of the tests (key already found vs timestamp), but that also makes sense and it's likely more performant as the dictionary will be empty initially and the first test will be fast, whereas comparing timestamps will take time.

Actually one simple optimization: check the size of `predecessorsByKey`. If it's the same as `keys`, then you can break out of the loop completely. For instance, when it's just one change, and thus just one key, there is no need to continue as soon as you find the first match.

One more optimization possible in the predicate: add a condition for the timestamp to be less than the highest timestamp from all the changes. Since you're likely running this only on a handful of changes, that means a smallest list of changes to go through.

**Comment by drewmccormack** [comment]
Thanks for checking this. 

Will consider the optimisations. The first one sounds good. An early return that could potentially save some time. Would depend on the 'randomness' of key changes, I guess. Otherwise you are probably going back to the original object anyway for some properties, and don't save much.

Second one could help, though not clear, because I would first have to calculate the maximum timestamp, and in most cases we use it, we are working at a recent time anyway, so unlikely to skim much off.

**Comment by cparnot** [comment]
Yeah, the first optimization is close to a no brainer in that checking for the number of elements should be fast. OTOH, the main bottleneck is probably I/O.

The second one might be better than you'd think if it reduces the amount of data to pull off the database, and impact I/O.

One clear optimization would be for single changes (or single key). Not sure if you expect many of those, but if you do, I would go as far as making a separate implementation for that special case, and call it from the general one.

btw, I realize some of that may be premature optimization, but mostly what I am looking at here, is things that are easy to understand and reason about, so easy to write... and read.

**Comment by cparnot** [comment]
Excellent. Changes looking good.

File: Core/PARStore.m [hunk 1]
```diff
@@ -2331,33 +2331,48 @@ - (NSArray *)fetchChangesFromTimestamp:(nullable NSNumber *)firstTimestamp toTim
...
+        if (versionsByKey[key]) continue;
         
-        // Change must preceed our final change
-        PARChange *finalChange = finalChangesByKey[key];
-        if (change.timestamp >= finalChange.timestamp) continue;
+        // Check version is on the right side of change in time
+        PARChange *version = finalChangesByKey[key];
+        BOOL versionIsFirst = change.timestamp > version.timestamp;
+        BOOL timesAreEqual = change.timestamp == version.timestamp;
+        if (timesAreEqual || (versionPreceeds ^ versionIsFirst)) continue;
```
Comment (cparnot) [inline Core/PARStore.m]: Nice. We don't use XOR often enough :-)

Comment (drewmccormack) [reply Core/PARStore.m]: Indeed, that's because it is the same as != in this case, which is what I should have used to begin with. Changing.

**Comment by drewmccormack** [comment]
OK, I'm on the verge of merging this into master, so worth taking a fresh look before I do.

**Pull request merged**